From: Jaroslav Kysela <perex.suse.cz>
Subject: Fix debug traces for bonding (might_sleep)
References: 174843

The network bonding module might call rtmsg_ifinfo in the atomic context.
It is not ideal solution - ideally, the bonding module will use a workqueue.

Signed-off-by: Jaroslav Kysela <perex@suse.cz>
Acked-by: Olaf Kirch <okir@suse.de>

--- linux-2.6.16.orig/net/core/rtnetlink.c	2006-03-20 06:53:29.000000000 +0100
+++ linux-2.6.16/net/core/rtnetlink.c	2006-05-23 17:25:00.000000000 +0200
@@ -451,7 +451,7 @@ void rtmsg_ifinfo(int type, struct net_d
 			       sizeof(struct rtnl_link_ifmap) +
 			       sizeof(struct rtnl_link_stats) + 128);
 
-	skb = alloc_skb(size, GFP_KERNEL);
+	skb = alloc_skb(size, in_atomic() ? GFP_ATOMIC : GFP_KERNEL);
 	if (!skb)
 		return;
 
@@ -460,7 +460,7 @@ void rtmsg_ifinfo(int type, struct net_d
 		return;
 	}
 	NETLINK_CB(skb).dst_group = RTNLGRP_LINK;
-	netlink_broadcast(rtnl, skb, 0, RTNLGRP_LINK, GFP_KERNEL);
+	netlink_broadcast(rtnl, skb, 0, RTNLGRP_LINK, in_atomic() ? GFP_ATOMIC : GFP_KERNEL);
 }
 
 /* Protected by RTNL sempahore.  */
